npm/webpack-preprocessor/test/unit/index.spec.js

```diff
@@ -91,7 +91,14 @@ describe('webpack preprocessor', function () {
       })
 
       it('runs webpack', function () {
+        expect(preprocessor.__bundles()[this.file.filePath]).to.be.undefined
+
         return this.run().then(() => {
+          // resolve prop should not exist, since promise should have been resolved.
```

Comment:
The `promise` should never have a `resolve` or `reject` property, regardless of its fulfillment, so if the goal is to test that the promise was resolved, we probably need to find a different way. Only the deferred objects have `resolve` and `reject` methods and they aren't removed upon fulfillment.

Reply:
I guess these are part of `Bluebird`? I've never seen the `resolve` or `reject` properties before.

I thought about this a bit more, but I don't see any good way to test these - I had to add a private ` __bundles` method to even access them, which feels like a code smell.

Do you have any suggestion on how we test this? Ideally we'd just add some kind of test project an exercise it in an e2e manner, but it's not even clear how to reproduce this in the first place.

Reply:
We [create the deferred object ourselves](https://github.com/cypress-io/cypress/blob/develop/npm/webpack-preprocessor/deferred.ts). Bluebird had an API for it, but it's deprecated and they [recommend doing it the way we are doing it](http://bluebirdjs.com/docs/api/deferred-migration.html).

I realize I said we need a different way to test the promise is resolved, but really it's not necessary because being in the `then` callback already proves that the promise resolved. And I don't think it's a sufficient test anyway. If it won't fail without the fix, it's not really testing the fix.

You're right that an e2e test might be best for this. We're currently only testing against webpack 4 as it is and we should test against webpack 5 as well. We do have [e2e tests](https://github.com/cypress-io/cypress/tree/37de5f8dc5ecbd72c3030f6880a7f74a56b953f8/npm/webpack-preprocessor/test/e2e) already. [This one](https://github.com/cypress-io/cypress/blob/37de5f8dc5ecbd72c3030f6880a7f74a56b953f8/npm/webpack-preprocessor/test/e2e/e2e.spec.js) does an entire Cypress run. We should be able to have both webpack 4 and 5 available ([similar to the way the driver does for react](https://github.com/cypress-io/cypress/blob/37de5f8dc5ecbd72c3030f6880a7f74a56b953f8/packages/driver/package.json#L65)) and then "install" the desired one before the e2e test [runs cypress](https://github.com/cypress-io/cypress/blob/37de5f8dc5ecbd72c3030f6880a7f74a56b953f8/npm/webpack-preprocessor/test/e2e/helpers.js#L39).

I don't know exactly how that will work, but I think it's possible. Let me know if you'd like to pair on it and we can schedule some time.

Reply:
```suggestion
```

Need to remove the comment too.